PhysicsTools/NanoAOD/plugins/DSAMuonTableProducer.cc

```diff
@@ -0,0 +1,440 @@
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+#include "DataFormats/NanoAOD/interface/FlatTable.h"
+
+#include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "TrackingTools/IPTools/interface/IPTools.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/VertexReco/interface/VertexFwd.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+
+class DSAMuonTableProducer : public edm::global::EDProducer<> {
+public:
+  DSAMuonTableProducer(const edm::ParameterSet& iConfig)
+      : dsaMuonTag_(consumes<std::vector<reco::Track>>(iConfig.getParameter<edm::InputTag>("dsaMuons"))),
+        muonTag_(consumes<std::vector<pat::Muon>>(iConfig.getParameter<edm::InputTag>("muons"))),
+        vtxTag_(consumes<reco::VertexCollection>(iConfig.getParameter<edm::InputTag>("primaryVertex"))),
+        bsTag_(consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("beamspot"))),
+        transientTrackBuilderToken_(esConsumes(edm::ESInputTag("", "TransientTrackBuilder"))) {
+    produces<nanoaod::FlatTable>("DSAMuon");
+  }
+
+  ~DSAMuonTableProducer() override {}
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<edm::InputTag>("dsaMuons")->setComment("input displaced standalone muon collection");
+    desc.add<edm::InputTag>("muons")->setComment("input muon collection");
+    desc.add<edm::InputTag>("primaryVertex")->setComment("input primary vertex collection");
+    desc.add<edm::InputTag>("beamspot")->setComment("input beamspot collection");
+    descriptions.add("DSAMuonTable", desc);
+  }
+
+private:
+  void produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const override;
+
+  bool passesDisplacedID(const reco::Track& dsaMuon) const;
+  int getMatches(const pat::Muon& muon, const reco::Track& dsaMuon, float minPositionDiff = 1e-6) const;
+  int getDTMatches(const pat::Muon& muon, const reco::Track& dsaMuon, float minPositionDiff = 1e-6) const;
+  int getCSCMatches(const pat::Muon& muon, const reco::Track& dsaMuon, float minPositionDiff = 1e-6) const;
```

Comment:
These three functions are essentially performing the same logic modulo minor differences. Such approach implies repeating almost-identical loops multiple times.

I would suggest writing a single function that accumulates on a `struct` containing three `int` reporting about _all matches_, _DT matches_ and _CSC matches_, and returning the struct.

Also you could even store in the `struct` just two `int` if _all matches_ is fully redundant.